test/integration-test/src/tests/ring_buf.rs

```diff
@@ -0,0 +1,440 @@
+
+use anyhow::Context as _;
+use assert_matches::assert_matches;
+use aya::{
+    maps::{array::PerCpuArray, ring_buf::RingBuf, MapData},
+    programs::UProbe,
+    Bpf, BpfLoader, Btf, Pod,
+};
+use aya_obj::generated::BPF_RINGBUF_HDR_SZ;
+use rand::Rng as _;
+use tokio::{
+    io::unix::AsyncFd,
+    time::{sleep, Duration},
+};
+
+// This structure's definition is duplicated in the probe.
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Default)]
+struct Registers {
+    dropped: u64,
+    rejected: u64,
+}
+
+impl std::ops::Add for Registers {
+    type Output = Self;
+    fn add(self, rhs: Self) -> Self::Output {
+        Self {
+            dropped: self.dropped + rhs.dropped,
+            rejected: self.rejected + rhs.rejected,
+        }
+    }
+}
+
+impl<'a> std::iter::Sum<&'a Registers> for Registers {
+    fn sum<I: Iterator<Item = &'a Registers>>(iter: I) -> Self {
+        iter.fold(Default::default(), |a, b| a + *b)
+    }
+}
+
+unsafe impl Pod for Registers {}
+
+struct RingBufTest {
+    _bpf: Bpf,
+    ring_buf: RingBuf<MapData>,
+    regs: PerCpuArray<MapData, Registers>,
+}
+
+// Note that it is important for this test that RING_BUF_MAX_ENTRIES ends up creating a ring buffer
+// that is exactly a power-of-two multiple of the page size. The synchronous test will fail if
+// that's not the case because the actual size will be rounded up, and fewer entries will be dropped
+// than expected.
+const RING_BUF_MAX_ENTRIES: usize = 512;
+
+impl RingBufTest {
+    fn new() -> Self {
+        const RING_BUF_BYTE_SIZE: u32 =
+            (RING_BUF_MAX_ENTRIES * (mem::size_of::<u64>() + BPF_RINGBUF_HDR_SZ as usize)) as u32;
+
+        // Use the loader API to control the size of the ring_buf.
+        let mut bpf = BpfLoader::new()
+            .btf(Some(Btf::from_sys_fs().unwrap()).as_ref())
+            .set_max_entries("RING_BUF", RING_BUF_BYTE_SIZE)
+            .load(crate::RING_BUF)
+            .unwrap();
+        let ring_buf = bpf.take_map("RING_BUF").unwrap();
+        let ring_buf = RingBuf::try_from(ring_buf).unwrap();
+        let regs = bpf.take_map("REGISTERS").unwrap();
+        let regs = PerCpuArray::<_, Registers>::try_from(regs).unwrap();
+        let prog: &mut UProbe = bpf
+            .program_mut("ring_buf_test")
+            .unwrap()
+            .try_into()
+            .unwrap();
+        prog.load().unwrap();
+        prog.attach(
+            Some("ring_buf_trigger_ebpf_program"),
+            0,
+            "/proc/self/exe",
+            None,
+        )
+        .unwrap();
+
+        Self {
+            _bpf: bpf,
+            ring_buf,
+            regs,
+        }
+    }
+}
+
+struct WithData(RingBufTest, Vec<u64>);
+
+impl WithData {
+    fn new() -> Self {
+        Self(RingBufTest::new(), {
+            let mut rng = rand::thread_rng();
+            // Generate more entries than there is space so we can test dropping entries.
+            let n = rng.gen_range(1..=RING_BUF_MAX_ENTRIES * 2);
```

Comment:
The comment doesn't match the actual value of n. This _sometimes_ will test
values over RING_BUF_MAX_ENTRIES. I don't think this kind of randomness here is
good? We should make sure both cases (in range, out of range) are always tested.

In general randomness in tests should be avoided. A test fails, I run it
again and I have new inputs, so it may or may not fail. In this case the random
values for the entries are probably fine, but imo `n` shouldn't be random here.

Reply:
I parameterized the tests on N and then added test cases so that the data sizes for the relevant tests are both under and over capacity.